Reusable Blocks: Add reusable blocks effects

noisysocks opened the pull request:
Adds the effects necessary for supporting reusable blocks. There are 4:

- FETCH_REUSABLE_BLOCKS: Loads reusable blocks from the API and inserts
  them into the store.
- SAVE_REUSABLE_BLOCK: Persists a reusable block that's in the store to
  the API.
- MAKE_BLOCK_STATIC: Transforms a reusable block on the page into a
  regular block.
- MAKE_BLOCK_REUSABLE: Transforms a regular block on the page into a
  reusable block.

noisysocks committed:
Add reusable blocks effects

Adds the effects necessary for supporting reusable blocks. There are 4:

- FETCH_REUSABLE_BLOCKS: Loads reusable blocks from the API and inserts
  them into the store.
- SAVE_REUSABLE_BLOCK: Persists a reusable block that's in the store to
  the API.
- MAKE_BLOCK_STATIC: Transforms a reusable block on the page into a
  regular block.
- MAKE_BLOCK_REUSABLE: Transforms a regular block on the page into a
  reusable block.

youknowriad commented on lib/client-assets.php:
```diff
@@ -752,6 +752,11 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
+	// Ensure that we've got jQuery.
+	if ( ! wp_script_is( 'jquery', 'done' ) ) {
+		wp_enqueue_script( 'jquery' );
+	}
```
Why can't we just add a `jquery` dependency, something like this https://github.com/WordPress/gutenberg/pull/3345/commits/255ef193d1c0a0bd30e1863dc89da4875f799b86

noisysocks replied:
We can only pass dependencies into `wp_enqueue_script`, not `wp_add_inline_script`: https://developer.wordpress.org/reference/functions/wp_add_inline_script/

aduth replied:
But `wp_add_inline_script` is tied to the original handle `wp-editor`, and we can define jQuery as a dependency for the enqueuing of `wp-editor`.

youknowriad commented on lib/client-assets.php:
```diff
@@ -765,9 +770,14 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
+	wp_add_inline_script(
+		'wp-editor',
+		(
+			'jQuery.when( wp.api.init(), wp.api.init( { versionString: \'gutenberg/v1/\' } ) ).done( function() {'
```
Is `jQuery.when` the same as `Promise.all`. It would be great if we can avoid `jQuery`

aduth replied:
I expect `init` may be returning an instance of [`jQuery.Deferred`](https://api.jquery.com/jquery.deferred/). Unsure if `Promise.all` is capable of handling these.

noisysocks replied:
It looks like that would work: https://codepen.io/noisysocks/pen/zPROqr?editors=0011

noisysocks replied:
👌 fixed in b2e1a7d054c06e26a2586a846dc4124292a27fc7.

noisysocks replied:
Wait, we can't do this though because IE 11 doesn't support promises! https://caniuse.com/#feat=promises

youknowriad replied:
We use a promise polyfill :) https://github.com/WordPress/gutenberg/blob/6eec1886b9270073ab7cf6ca2543ad39640dafb1/lib/client-assets.php#L141

noisysocks replied:
Right you are! 😄 Changed in f5cd6cded675b9dec31d52ec961064f7a060fa04.

youknowriad commented on lib/client-assets.php:
```diff
@@ -765,9 +770,14 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
+		'wp-editor',
+		(
+			'jQuery.when( wp.api.init(), wp.api.init( { versionString: \'gutenberg/v1/\' } ) ).done( function() {'
+			. 'wp.editor.createEditorInstance( \'editor\', window._wpGutenbergPost, ' . json_encode( $editor_settings ) . ' ); '
```
I guess we should drop this line? Probably a merge conflict?

noisysocks replied:
Oops! Fixed in 5367b6e970cbb7e93c97a28f61459f7ffdee2f25.

noisysocks committed:
Remove extraneous createEditorInstance

This unnecessary line probably came as a result of a bad merge conflict
resolution.

noisysocks committed:
Fix 'cannot call initializeMetaBoxes on undefined' error

`gutenberg_collect_meta_box_data` expected that
`window._wpGutenbergEditor` would always be instantiated, but this is
only true for when `wp.api.init()` returns quickly because of the API
schema being cached.

The fix is to use a Promise to store `window._wpGutenbergEditor`.

(Well, a jQuery.Deferred, since we have to support IE 11 which doesn't
support promises.)

noisysocks committed:
Replace jQuery.Deferred with Promise

Remove jQuery.Deferred and jQuery.when in our editor initialization in
lieu of Promise and Promise.all.

youknowriad commented on lib/register.php:
```diff
@@ -236,7 +236,7 @@ function gutenberg_collect_meta_box_data() {
 	wp_add_inline_script(
 		'wp-editor',
-		'window._wpGutenbergEditor.initializeMetaBoxes( ' . wp_json_encode( $meta_box_data ) . ' )'
+		'window._wpGutenbergEditor.then( function( editor ) { editor.initializeMetaBoxes( ' . wp_json_encode( $meta_box_data ) . ' ) } );'
```
Good catch here! Potential bug fixed. Maybe we should rename the variable to make it clear that it's a promise and not the editor instance. Something like `_wpLoadGutenbergEditor`

noisysocks replied:
👍 e156849c16f4f75a1f39647b91214b55059f5ff5

youknowriad reviewed (approved):
Nice work

noisysocks committed:
Rename _wpGutenbergEditor -> _wpLoadGutenbergEditor

Make it clear that this variable is a promise by giving it a non-noun
name.
